sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -447,6 +453,12 @@ class ArtifactManager(session: SparkSession) extends Logging {
     }
     fs.copyFromLocalFile(false, true, new FSPath(localPath.toString), destFSPath)
   }
+
+  // The Cleaner and the associated cleanup task
+  private val cleaner: Cleaner = Cleaner.create()
+  private val cleanable = cleaner.register(this, new Runnable {
```

Comment:
This might be an issue... The runnable you create here points to the class you are trying to clean-up. This will keep it alive. AFAIK we will have to separate the state from the ArtifactManager to make this work.

Reply:
Ah oof, good point...